src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -419,4 +418,128 @@ public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable Index
             }
         }
     }
+
+    /**
+     * Deletes the given index. Persistent parts of the index
+     * like the shards files, state and transaction logs are removed once all resources are released.
+     *
+     * Equivalent to {@link #removeIndex(String, String)} but fires
+     * different lifecycle events to ensure pending resources of this index are immediately removed.
+     * @param index the index to delete
+     * @param reason the high level reason causing this delete
+     */
+    public void deleteIndex(String index, String reason) throws IOException {
+        removeIndex(index, reason, true);
+    }
+
+    public void deleteClosedIndex(String reason, ClusterState clusterState, IndexMetaData metaData) {
+        try {
+            deleteIndexStore(reason, clusterState, metaData);
+        } catch (IOException e) {
+            logger.warn("[{}] failed to delete closed index", e, metaData.index());
+        }
+    }
+
+    /**
+     * Deletes the index store trying to acquire all shards locks for this index.
+     * This method will delete the metadata for the index even if the actual shards can't be locked.
+     */
+    public void deleteIndexStore(String reason, ClusterState clusterState, IndexMetaData metaData) throws IOException {
+        if (nodeEnv.hasNodeFile()) {
+            String indexName = metaData.getIndex();
+            if (clusterState.metaData().hasIndex(indexName)) {
+                final IndexMetaData index = clusterState.metaData().index(indexName);
+                throw new ElasticsearchIllegalStateException("Can't delete index store for [" + indexName + "] - it's still part of the cluster state [" + index.getUUID() + "] [" + metaData.getUUID() + "]");
+            }
+            Index index = new Index(indexName);
+            final Settings indexSettings = buildIndexSettings(metaData);
+            deleteIndexStore(reason, index, indexSettings);
+        }
+    }
+
+    private void deleteIndexStore(String reason, Index index, Settings indexSettings) throws IOException {
+        try {
+            // we are trying to delete the index store here - not a big deal if the lock can't be obtained
+            // the store metadata gets wiped anyway even without the lock this is just best effort since
+            // every shards deletes its content under the shard lock it owns.
+            logger.info("deleting index store reason [{}]", reason);
```

Comment:
I think this should be a trace log with "[{}] trying to delete store (reason: [{}})" (note the index name as well) and log the info after successful deletion. We already log a debug message (or higher) in all callers of this method and it will not result in an info message saying "deleting" when we actually failed.

```
            logger.trace("{} trying to deleting store (reason [{}]}", index, reason);
            nodeEnv.deleteIndexDirectorySafe(index, 0, indexSettings);
            logger.info("{} store deleted (reason [{}]}", index, reason);
        } catch (LockObtainFailedException ex) {
            logger.debug("{} failed to delete index store - at least one shards is still locked", ex, index);

```